pkg/readiness/pruner/pruner_test.go

```diff
@@ -0,0 +1,185 @@
+
+	events := make(chan event.GenericEvent, 1024)
+	reg, err := wm.NewRegistrar(
+		cachemanager.RegistrarName,
+		events,
+	)
+	require.NoError(t, err, "creating registrar")
+
+	cfClient := &fakes.FakeCfClient{}
+	config := &cachemanager.Config{
+		CfClient:         cfClient,
+		SyncMetricsCache: syncutil.NewMetricsCache(),
+		Tracker:          tracker,
+		ProcessExcluder:  process.Get(),
+		Registrar:        reg,
+		Reader:           c,
+		GVKAggregator:    aggregator.NewGVKAggregator(),
+	}
+	require.NoError(t, err, "creating registrar")
+	cm, err := cachemanager.NewCacheManager(config)
+	require.NoError(t, err, "creating cachemanager")
+
+	if !startControllers {
+		// need to start the cachemanager if controllers are not started
+		// since the cachemanager is started in the controllers code.
+		require.NoError(t, mgr.Add(cm), "adding cachemanager as a runnable")
+	} else {
+		sw := watch.NewSwitch()
+		mutationSystem := mutation.NewSystem(mutation.SystemOpts{})
+
+		frameworksexternaldata.NewCache()
+		opts := controller.Dependencies{
+			CFClient:         testutils.SetupDataClient(t),
+			WatchManger:      wm,
+			ControllerSwitch: sw,
+			Tracker:          tracker,
+			ProcessExcluder:  process.Get(),
+			MutationSystem:   mutationSystem,
+			ExpansionSystem:  expansion.NewSystem(mutationSystem),
+			ProviderCache:    frameworksexternaldata.NewCache(),
+			CacheMgr:         cm,
+			SyncEventsCh:     events,
+		}
+		require.NoError(t, controller.AddToManager(mgr, &opts), "registering controllers")
+	}
+
+	testutils.StartManager(ctx, t, mgr)
+
+	return &ExpectationsPruner{
+		cacheMgr: cm,
+		tracker:  tracker,
+	}, c
+}
+
+// Test_ExpectationsMgr_DeletedSyncSets tests scenarios in which SyncSet and Config resources
+// get deleted after tracker expectations have been populated and we need to reconcile
+// the GVKs that are in the data client (via the cachemaanger) and the GVKs that are expected
+// by the Tracker.
+func Test_ExpectationsMgr_DeletedSyncSets(t *testing.T) {
+	tts := []struct {
+		name             string
+		fixturesPath     string
+		syncsetsToDelete []string
+		deleteConfig     string
+		startControllers bool
+	}{
+		{
+			name:             "delete all syncsets",
+			fixturesPath:     "testdata/syncsets-overlapping",
+			syncsetsToDelete: []string{"syncset-1", "syncset-2", "syncset-3"},
+		},
+		{
+			name:             "delete syncs and configs",
+			fixturesPath:     "testdata/syncsets-config-disjoint",
+			syncsetsToDelete: []string{"syncset-1"},
+			deleteConfig:     "config",
+		},
+		{
+			name:             "delete one syncset",
+			fixturesPath:     "testdata/syncsets-resources",
+			syncsetsToDelete: []string{"syncset-2"},
+			startControllers: true,
+		},
+	}
+
+	for _, tt := range tts {
+		t.Run(tt.name, func(t *testing.T) {
+			ctx, cancelFunc := context.WithCancel(context.Background())
+
+			require.NoError(t, testutils.ApplyFixtures(tt.fixturesPath, cfg), "applying base fixtures")
```

Comment:
I don't understand how `ApplyFixtures` works.  Dont' we need to fill a fake client with data and then return that fake client?  That fake client is then full of objects, acting as a dummy API server.  How does this one work?

Reply:
We don't really pass along a fake client in some of our other tests either. I think the common kernel to all test types is having that same `cfg *rest.Config` be used to construct clients that point to the same fake control plane. In this case, we create one client for the given rest config and add our fixtures to it. Then, when we create a new client out of the manager in `setupTest`,but it's all the same control plane in the background that we used for ApplyFixtures. Does it make sense ?